draco/analysis/flagging.py

```diff
@@ -1755,3 +1674,207 @@ def process(self, data):
+        """
+
+        self.telescope = io.get_telescope(telescope)
+
+    def process(
+        self,
+        stream: Union[
+            containers.TimeStream, containers.SiderealStream, andata.CorrData
+        ],
+    ) -> Union[containers.BaselineMask, containers.SiderealBaselineMask]:
+        """Construct mask and apply to input weights, if desired.
+
+        Parameters
+        ----------
+        stream : `.core.container` with `weight` attribute
+            Input container whose weights are used to construct the mask.
+
+        Returns
+        -------
+        out : `BaselineMask` or `SiderealBaselineMask`
+            The output baseline-dependent mask.
+        """
+        from mpi4py import MPI
+
+        # Only redistribute the weight dataset, because CorrData containers will have
+        # other parallel datasets without a stack axis
+        stream.weight.redistribute(axis=1)
+
+        # Make the output container, depending on input type
+        if "ra" in stream.axes:
+            mask_cont = containers.SiderealBaselineMask(
+                axes_from=stream, attrs_from=stream
+            )
+        elif "time" in stream.axes:
+            mask_cont = containers.BaselineMask(axes_from=stream, attrs_from=stream)
+        else:
+            raise TypeError(
+                f"Task requires TimeStream, SiderealStream, or CorrData. Got {type(stream)}"
+            )
+
+        # Redistribute output container along stack axis
+        mask_cont.redistribute("stack")
+
+        # Get local section of weights
+        local_weight = stream.weight[:].local_array
+
+        # Define averaging function that takes mean or median of nonzero elements.
+        # If all elements are zero, simply return zero.
+        average_func = np.median if self.average_type == "median" else np.mean
+        average_nonzero_func = (
+            lambda x: average_func(x[x != 0].astype(float))
+            if np.sum(x != 0) > 0
+            else 0.0
+        )
+
+        # Apply averaging function to weights for each baseline
+        average_weight = np.array(
+            [average_nonzero_func(w) for w in local_weight.transpose(1, 0, 2)],
+            dtype=local_weight.dtype,
+        )
+
+        # Figure out which entries to keep
+        threshold = np.maximum(
+            self.absolute_threshold, self.relative_threshold * average_weight
+        )
+
+        # Compute the mask, excluding samples that already have zero weight
+        local_mask = (local_weight < threshold[np.newaxis, :, np.newaxis]) & (
+            local_weight != 0
+        )
+
+        # If only flagging co-pol baselines, make separate mask to select those,
+        # and multiply into low-weight mask
+        if self.pols_to_flag == "copol":
+
+            # Get local section of stack axis
+            stack_lo = stream.weight.local_offset[1]
+            stack_ls = stream.weight.local_shape[1]
+            local_stack = stream.stack[stack_lo : stack_lo + stack_ls]
```

Comment:
```suggestion
            local_stack = stream.stack[local_weight.local_bounds]
```
MPIArray.local_bounds is a fairly new method to simplify getting a slice like this

Reply:
I had to use `stream.weight[:].local bounds`, but otherwise it works out

Reply:
Right, yeah the local_weight would already by a numpy array